app-header/app-header.html

```diff
@@ -45,184 +92,324 @@
-      },
+      // prepare for measurement
+      this.noTransitions = true;
 
-      _fixedChanged: function(fixed) {
-        if (fixed) {
-          this._transformHeader(0);
-        }
+      if (firstSetup || scrollTop === 0) {
+        this.cancelDebouncer('_resize');
+        this._deltaHeight = this._topbar ? this._height - this._topbar.offsetHeight : 0;
+        this._setUpEffect();
+        this._updateScrollState(scrollTop, true);
+      } else {
+        this._updateScrollState(0, true);
+        this._deltaHeight = this._topbar ? this._height - this._topbar.offsetHeight : 0;
+        this._setUpEffect();
+        this._updateScrollState(this._lastScrollTop, true);
+        this._layoutIfDirty();
       }
+      // restore no transition
+      this.noTransitions = currentNoTransition;
+    },
+
+    /**
+     * Updates the scroll state.
+     *
+     * @param {number}
+     * @param {boolean}
+     */
+    _updateScrollState: function(scrollTop, forceUpdate) {
+      if (this._height === 0) {
+        return;
+      }
+      var progress;
+      var mayHeaderMove = !this.fixed || this.condenses;
+      var y = Math.min(this._headerDy, Math.max(0,
+          this.reveals ? this._y + scrollTop - this._lastScrollTop : scrollTop
+      ));
+
+      if (this._shouldCondense(scrollTop)) {
+        y = Math.max(this._deltaHeight, y);
+      }
+      if (this._deltaHeight === 0) {
+        progress = scrollTop > 0 ? 1 : 0;
+      } else {
+        progress = y / this._deltaHeight;
+      }
+      if (forceUpdate || progress !== this._progress || y !== this._y || scrollTop === 0) {
+        this._runEffects(progress, mayHeaderMove ? y : 0);
+        this._transformHeader(mayHeaderMove ? y : 0);
+        this.fire('app-header-transform', { y: y, progress: progress }, { bubbles: false });
+      }
+      if (!forceUpdate) {
+        this._y = y;
+        this._lastScrollTop = scrollTop;
+        this._progress = progress;
+      }
+    },
+
+    /**
+     * Returns true if the current header can condense.
+     *
+     * @param {number} scrollTop
+     * @return {boolean}
+     */
+    _shouldCondense: function(scrollTop) {
+      return this._lastScrollTop >= scrollTop && scrollTop > this._deltaHeight
+          && (!this.reveals || this.condenses);
+    },
+
+    /**
+     * Returns true if the current element is on the screen.
+     * That is, visible in the current viewport. This method should be
+     * overridden by the consumer of this behavior.
+     *
+     * @method isOnScreen
+     * @return {boolean}
+     */
+    isOnScreen: function() {
+      return this._height !== 0 && (this._y === 0 || this._y < this._height);
+    },
+
+    /**
+     * Returns true if there's content below the current element. This method
+     * should be overridden by the consumer of this behavior.
+     *
+     * @method isContentBelow
+     * @return {boolean}
+     */
+    isContentBelow: function(headerY) {
+      if (headerY === 0) {
```

Comment:
In the doc, we should explain what `headerY` is so someone can override it if they want to.